sql/src/main/java/com/rilldata/SqlConverter.java

```diff
@@ -1,34 +1,51 @@
 package com.rilldata;
 
 import com.rilldata.calcite.CalciteToolbox;
+import com.rilldata.calcite.models.SqlCreateMetricsView;
+import com.rilldata.calcite.models.SqlCreateSource;
 import org.apache.calcite.sql.SqlDialect;
-import org.apache.calcite.sql.dialect.PostgresqlSqlDialect;
-
-import javax.sql.DataSource;
-import java.sql.SQLException;
+import org.apache.calcite.sql.SqlLiteral;
+import org.apache.calcite.sql.parser.SqlParseException;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.tools.ValidationException;
 
 public class SqlConverter
 {
-  private DataSource datasource;
   private CalciteToolbox calciteToolbox;
 
-  public SqlConverter(String schema) throws SQLException
+  public SqlConverter(String schema)
   {
     calciteToolbox = new CalciteToolbox(new StaticSchemaProvider(schema), null);
   }
 
-  public void initialize(String ddl) throws SQLException
+  public String convert(String sql, SqlDialect sqlDialect) throws ValidationException, SqlParseException
   {
+    return calciteToolbox.getRunnableQuery(sql, sqlDialect);
   }
 
-  public String convert(String sql, SqlDialect sqlDialect)
+  public byte[] createSource(String sourceDef)
   {
     try {
-      return calciteToolbox.getRunnableQuery(sql, sqlDialect);
+      SqlCreateSource sqlCreateSource = calciteToolbox.createSource(sourceDef);
+      return calciteToolbox.getAST(sqlCreateSource);
+    } catch (Exception e) {
+      e.printStackTrace(); // todo level-logging for native libraries?
+      // in case of error returning an AST containing StringLiteral with error messages as the top most node
+      return calciteToolbox.getAST(
+          SqlLiteral.createCharString(String.format("{'error': '%s'}", e.getMessage()), new SqlParserPos(0, 0)));
```

Comment:
Yes, thats what the code does. This line actually creates a SqlLiteral with error message like `{'error': '%s'}` so at the runtime code needs to check if the returned AST has a SqlLiteral as the top most node the literal is a json with error. We can actually add one `message` to protobuf just for the error which is not a SqlNode to make it more clearer and easier to read at runtime side.

Reply:
Yeah I saw the SqlLiteral with error messages. It's fine for short term. Instead of overloading the AST to carry error messages, I think it might be better to have a separate `.proto` (e.g.`rpc.proto`) file declaring the RPC interface. Protobuf files can actually import each other, so the RPC interface can have SQL ASTs as subtypes.